File: server-csharp/Collisions/EPA/Methods.cs
using System.Diagnostics.Contracts;
using System.Numerics;
using SpacetimeDB;

    public static partial class Module
    {
        public static bool EpaSolve(
        GjkResult Gjk,
        List<ConvexHullCollider> ColliderA,
        DbVector3 PositionA,
        float YawRadiansA,
        List<ConvexHullCollider> ColliderB,
        DbVector3 PositionB,
        float YawRadiansB,
        out Contact Contact)
    {
        const int MaxIterations = 32;
        const float Epsilon = 5e-4f;

        Contact = new Contact
        {
            Normal = new DbVector3(0f, 1f, 0f),
            Depth = 0f
        };

        if (Gjk.Simplex == null || Gjk.Simplex.Count < 4)
        {
            return false;
        }

        List<GjkVertex> PolytopeVertices = new List<GjkVertex>(Gjk.Simplex);

        List<EpaFace> Faces = new List<EpaFace>();
        AddFace(PolytopeVertices, Faces, 0, 1, 2);
        AddFace(PolytopeVertices, Faces, 0, 3, 1);
        AddFace(PolytopeVertices, Faces, 0, 2, 3);
        AddFace(PolytopeVertices, Faces, 1, 3, 2);

        for (int Iteration = 0; Iteration < MaxIterations; Iteration++)
        {
            int ClosestFaceIndex = -1;
            float ClosestDistance = float.MaxValue;

            for (int FaceIndex = 0; FaceIndex < Faces.Count; FaceIndex++)
            {
                EpaFace Face = Faces[FaceIndex];
                if (Face.Obsolete)
                {
                    continue;
                }

                if (Face.Distance < ClosestDistance)
                {
                    ClosestDistance = Face.Distance;
                    ClosestFaceIndex = FaceIndex;
                }
            }

            if (ClosestFaceIndex < 0)
            {
                return false;
            }

            EpaFace ClosestFace = Faces[ClosestFaceIndex];
            DbVector3 SearchDirection = ClosestFace.Normal;

            GjkVertex NewVertex = SupportPairWorld(
                ColliderA, PositionA, YawRadiansA,
                ColliderB, PositionB, YawRadiansB,
                SearchDirection);

            float Projection = Dot(NewVertex.MinkowskiPoint, SearchDirection);
            float Improvement = Projection - ClosestFace.Distance;

            if (Improvement < Epsilon)
            {
                DbVector3 Normal = ClosestFace.Normal;
                float NormalLengthSq = Dot(Normal, Normal);
                if (NormalLengthSq > 1e-12f)
                {
                    Normal = Mul(Normal, 1f / Sqrt(NormalLengthSq));
                }
                else
                {
                    Normal = NormalizeSmallVector(Gjk.LastDirection, new DbVector3(0f, 1f, 0f));
                }

                float Depth = ClosestFace.Distance;
                if (Depth < 0f)
                {
                    Depth = 0f;
                }

                DbVector3 RelativeBToA = Sub(PositionA, PositionB);
                if (Dot(Normal, RelativeBToA) < 0f)
                {
                    Normal = Negate(Normal);
                }

                Contact = new Contact
                {
                    Normal = Normal,
                    Depth = Depth
                };
                return true;
            }

            int NewVertexIndex = PolytopeVertices.Count;
            PolytopeVertices.Add(NewVertex);

            List<EpaEdge> Edges = new List<EpaEdge>();

            for (int FaceIndex = 0; FaceIndex < Faces.Count; FaceIndex++)
            {
                EpaFace Face = Faces[FaceIndex];
                if (Face.Obsolete)
                {
                    continue;
                }

                DbVector3 FacePoint = PolytopeVertices[Face.IndexA].MinkowskiPoint;
                DbVector3 ToNewPoint = Sub(NewVertex.MinkowskiPoint, FacePoint);
                float DotValue = Dot(Face.Normal, ToNewPoint);

                if (DotValue > 0f)
                {
                    Face.Obsolete = true;
                    Faces[FaceIndex] = Face;

                    AddEdge(Edges, Face.IndexA, Face.IndexB);
                    AddEdge(Edges, Face.IndexB, Face.IndexC);
                    AddEdge(Edges, Face.IndexC, Face.IndexA);
                }
            }

            for (int FaceIndex = Faces.Count - 1; FaceIndex >= 0; FaceIndex--)
            {
                if (Faces[FaceIndex].Obsolete)
                {
                    Faces.RemoveAt(FaceIndex);
                }
            }

            for (int EdgeIndex = 0; EdgeIndex < Edges.Count; EdgeIndex++)
            {
                EpaEdge Edge = Edges[EdgeIndex];
                if (Edge.Obsolete)
                {
                    continue;
                }

                AddFace(PolytopeVertices, Faces, Edge.IndexA, Edge.IndexB, NewVertexIndex);
            }
        }

        int FinalClosestFaceIndex = -1;
        float FinalClosestDistance = float.MaxValue;

        for (int FaceIndex = 0; FaceIndex < Faces.Count; FaceIndex++)
        {
            EpaFace Face = Faces[FaceIndex];
            if (Face.Obsolete)
            {
                continue;
            }

            if (Face.Distance < FinalClosestDistance)
            {
                FinalClosestDistance = Face.Distance;
                FinalClosestFaceIndex = FaceIndex;
            }
        }

        if (FinalClosestFaceIndex >= 0)
        {
            EpaFace Face = Faces[FinalClosestFaceIndex];

            DbVector3 Normal = Face.Normal;
            float NormalLengthSq = Dot(Normal, Normal);
            if (NormalLengthSq > 1e-12f)
            {
                Normal = Mul(Normal, 1f / Sqrt(NormalLengthSq));
            }
            else
            {
                Normal = NormalizeSmallVector(Gjk.LastDirection, new DbVector3(0f, 1f, 0f));
            }

            float Depth = Face.Distance;
            if (Depth < 0f)
            {
                Depth = 0f;
            }

            DbVector3 RelativeBToA = Sub(PositionA, PositionB);
            if (Dot(Normal, RelativeBToA) < 0f)
            {
                Normal = Negate(Normal);
            }

            Contact = new Contact
            {
                Normal = Normal,
                Depth = Depth
            };
            return false;
        }

        return false;
    }



    static void AddFace(List<GjkVertex> Vertices, List<EpaFace> Faces, int IndexA, int IndexB, int IndexC)
    {
        DbVector3 PointA = Vertices[IndexA].MinkowskiPoint;
        DbVector3 PointB = Vertices[IndexB].MinkowskiPoint;
        DbVector3 PointC = Vertices[IndexC].MinkowskiPoint;

        DbVector3 EdgeAB = Sub(PointB, PointA);
        DbVector3 EdgeAC = Sub(PointC, PointA);
        DbVector3 Normal = Cross(EdgeAB, EdgeAC);

        float LengthSquared = Dot(Normal, Normal);
        if (LengthSquared > 1e-12f)
        {
            float InverseLength = 1f / Sqrt(LengthSquared);
            Normal = Mul(Normal, InverseLength);
        }
        else
        {
            Normal = NormalizeSmallVector(Sub(PointB, PointC), new DbVector3(0f, 1f, 0f));
        }

        float Distance = Dot(Normal, PointA);
        if (Distance < 0f)
        {
            Normal = Negate(Normal);
            Distance = -Distance;
            (IndexC, IndexB) = (IndexB, IndexC);
        }

        EpaFace Face;
        Face.IndexA = IndexA;
        Face.IndexB = IndexB;
        Face.IndexC = IndexC;
        Face.Normal = Normal;
        Face.Distance = Distance;
        Face.Obsolete = false;

        Faces.Add(Face);
    }

    static void AddEdge(List<EpaEdge> Edges, int IndexA, int IndexB)
    {
        for (int EdgeIndex = 0; EdgeIndex < Edges.Count; EdgeIndex++)
        {
            EpaEdge ExistingEdge = Edges[EdgeIndex];
            if (!ExistingEdge.Obsolete && ExistingEdge.IndexA == IndexB && ExistingEdge.IndexB == IndexA)
            {
                ExistingEdge.Obsolete = true;
                Edges[EdgeIndex] = ExistingEdge;
                return;
            }
        }

        EpaEdge Edge;
        Edge.IndexA = IndexA;
        Edge.IndexB = IndexB;
        Edge.Obsolete = false;
        Edges.Add(Edge);
    }

    static DbVector3 ComputeContactNormal(DbVector3 RawNormal, DbVector3 CenterA, DbVector3 CenterB)
    {
        DbVector3 Normal = RawNormal;
        if (Dot(Normal, Normal) < 1e-6f) return new DbVector3(0f, 1f, 0f);
        Normal = Normalize(Normal);

        DbVector3 CenterDelta = Sub(CenterA, CenterB);
        float CenterDeltaSq = Dot(CenterDelta, CenterDelta);

        if (CenterDeltaSq > 1e-8f)
        {
            if (Dot(Normal, CenterDelta) < 0f)
                Normal = Negate(Normal);
        }

        DbVector3 WorldUp = new(0f, 1f, 0f);
        float UpDot = Dot(Normal, WorldUp);

        float FloorSnapDot = 0.98f;   // ~11 degrees of up
        float CeilingSnapDot = 0.98f; // ~11 degrees of down
        float WallSnapAbsDot = 0.05f; // ~87-93 degrees of up (horizontal)

        // FLOOR SNAP (nearly flat)
        if (UpDot >= FloorSnapDot)
        {
            return WorldUp;
        }

        // CEILING SNAP (nearly flat but inverted)
        if (UpDot <= -CeilingSnapDot)
        {
            return new DbVector3(0f, -1f, 0f);
        }

        // WALL SNAP (nearly vertical surface contact)
        if (MathF.Abs(UpDot) <= WallSnapAbsDot)
        {
            Normal.y = 0f;
            if (Dot(Normal, Normal) < 1e-6f) return new DbVector3(0f, 1f, 0f);
            return Normalize(Normal);
        }

        // SLOPE / RAMP (keep true normal)
        return Normal;
    }

}

File: server-rust/spacetimedb/src/Collisions/EPA/Methods.rs
use crate::*;

pub fn epa_solve(gjk: &GjkResult, collider_a: &Vec<ConvexHullCollider>, position_a: DbVector3, yaw_radians_a: f32, collider_b: &Vec<ConvexHullCollider>, position_b: DbVector3, yaw_radians_b: f32, contact_out: &mut Contact) -> bool { // Returns accurate collision normal and depth for two colliding objects
    let max_iterations: i32 = 16;
    let epsilon: f32 = 2e-3;
    let max_polytope_vertices: usize = 64;

    *contact_out = Contact { normal: DbVector3 { x: 0.0, y: 1.0, z: 0.0 }, depth: 0.0 };

    if gjk.simplex.len() < 4 { return false; }

    let mut polytope_vertices: Vec<GjkVertex> = gjk.simplex.clone();

    let mut faces: Vec<EpaFace> = Vec::new();
    faces.reserve(32);

    add_face(&polytope_vertices, &mut faces, 0, 1, 2);
    add_face(&polytope_vertices, &mut faces, 0, 3, 1);
    add_face(&polytope_vertices, &mut faces, 0, 2, 3);
    add_face(&polytope_vertices, &mut faces, 1, 3, 2);

    let mut edges: Vec<(i32, i32)> = Vec::new();
    edges.reserve(64);

    for _iteration in 0..max_iterations {
        let mut closest_face_index: i32 = -1;
        let mut closest_distance: f32 = f32::MAX;

        for face_index in 0..faces.len() {
            let face = &faces[face_index];
            if face.obsolete { continue; }
            if face.distance < closest_distance {
                closest_distance = face.distance;
                closest_face_index = face_index as i32;
            }
        }

        if closest_face_index < 0 { return false; }

        let search_direction = faces[closest_face_index as usize].normal;

        let new_vertex = support_pair_world(collider_a, position_a, yaw_radians_a, collider_b, position_b, yaw_radians_b, search_direction);
        let projection: f32 = dot(new_vertex.minkowski_point, search_direction);
        let improvement: f32 = projection - closest_distance;

        if improvement < epsilon {
            let mut normal = search_direction;
            let relative_b_to_a = sub(position_a, position_b);
            if dot(normal, relative_b_to_a) < 0.0 { normal = negate(normal); }

            let mut depth: f32 = closest_distance;
            if depth < 0.0 { depth = 0.0; }

            *contact_out = Contact { normal, depth };
            return true;
        }

        if polytope_vertices.len() >= max_polytope_vertices {
            let mut normal = search_direction;
            let relative_b_to_a = sub(position_a, position_b);
            if dot(normal, relative_b_to_a) < 0.0 { normal = negate(normal); }

            let mut depth: f32 = closest_distance;
            if depth < 0.0 { depth = 0.0; }

            *contact_out = Contact { normal, depth };
            return false;
        }

        let new_vertex_index: i32 = polytope_vertices.len() as i32;
        polytope_vertices.push(new_vertex);

        edges.clear();

        for face_index in 0..faces.len() {
            if faces[face_index].obsolete { continue; }

            let index_a = faces[face_index].index_a;
            let face_point = polytope_vertices[index_a as usize].minkowski_point;
            let to_new_point = sub(polytope_vertices[new_vertex_index as usize].minkowski_point, face_point);

            if dot(faces[face_index].normal, to_new_point) > 0.0 {
                faces[face_index].obsolete = true;

                let index_b = faces[face_index].index_b;
                let index_c = faces[face_index].index_c;

                add_edge_pair(&mut edges, index_a, index_b);
                add_edge_pair(&mut edges, index_b, index_c);
                add_edge_pair(&mut edges, index_c, index_a);
            }
        }

        let mut face_index: usize = 0;
        while face_index < faces.len() {
            if faces[face_index].obsolete {
                faces.swap_remove(face_index);
            } else {
                face_index += 1;
            }
        }

        for edge_index in 0..edges.len() {
            let (index_a, index_b) = edges[edge_index];
            add_face(&polytope_vertices, &mut faces, index_a, index_b, new_vertex_index);
        }
    }

    let mut final_closest_face_index: i32 = -1;
    let mut final_closest_distance: f32 = f32::MAX;

    for face_index in 0..faces.len() {
        let face = &faces[face_index];
        if face.obsolete { continue; }
        if face.distance < final_closest_distance {
            final_closest_distance = face.distance;
            final_closest_face_index = face_index as i32;
        }
    }

    if final_closest_face_index >= 0 {
        let mut normal = faces[final_closest_face_index as usize].normal;

        let relative_b_to_a = sub(position_a, position_b);
        if dot(normal, relative_b_to_a) < 0.0 { normal = negate(normal); }

        let mut depth: f32 = final_closest_distance;
        if depth < 0.0 { depth = 0.0; }

        *contact_out = Contact { normal, depth };
        return false;
    }

    false
}

pub fn add_face(vertices: &Vec<GjkVertex>, faces: &mut Vec<EpaFace>, index_a: i32, index_b: i32, index_c: i32) {
    let point_a = vertices[index_a as usize].minkowski_point;
    let point_b = vertices[index_b as usize].minkowski_point;
    let point_c = vertices[index_c as usize].minkowski_point;

    let edge_ab = sub(point_b, point_a);
    let edge_ac = sub(point_c, point_a);
    let mut normal = cross(edge_ab, edge_ac);

    let length_squared: f32 = dot(normal, normal);

    if length_squared > 1e-12 {
        normal = mul(normal, 1.0 / length_squared.sqrt());
    } else {
        normal = normalize_small_vector(sub(point_b, point_c), DbVector3 { x: 0.0, y: 1.0, z: 0.0 });
    }

    let mut distance: f32 = dot(normal, point_a);
    let mut final_index_b = index_b;
    let mut final_index_c = index_c;

    if distance < 0.0 {
        normal = negate(normal);
        distance = -distance;
        let temp = final_index_b;
        final_index_b = final_index_c;
        final_index_c = temp;
    }

    faces.push(EpaFace { index_a, index_b: final_index_b, index_c: final_index_c, normal, distance, obsolete: false });
}

pub fn add_edge_pair(edges: &mut Vec<(i32, i32)>, index_a: i32, index_b: i32) {
    let mut edge_index: usize = 0;
    while edge_index < edges.len() {
        let (existing_a, existing_b) = edges[edge_index];
        if existing_a == index_b && existing_b == index_a {
            edges.swap_remove(edge_index);
            return;
        }
        edge_index += 1;
    }

    edges.push((index_a, index_b));
}


